### PR TITLE
Refactor Jetpack to use the Asset_Tools package instead of Jetpack for get_file_url_for_environment

### PR DESCRIPTION
**Pull request by gititon**
This PR refactors Jetpack to use the Assets package for calling `get_file_url_for_environment` rather than the static method on the Jetpack class.

Since this is a refactor that should not introduce new functionality, please test by ensuring tests pass and by accessing various Jetpack pages/features and ensuring there are no errors in your error log.

In this latest revision, I used constructor dependency injection to inject Assets into a class when I could, and used a static get_instance() when it was needed from a static method.

### CODE REVIEW

File: class.jetpack-connection-banner.php [hunk 1]
```diff
@@ -1,6 +1,7 @@
 <?php
 
 use Automattic\Jetpack\Assets\Logo;
+use Automattic\Jetpack\Asset_Tools;
```
Comment (lezama) [inline class.jetpack-connection-banner.php]: should we go `use Automattic\Jetpack\Assets\Tools;` ?

Comment (lezama) [reply class.jetpack-connection-banner.php]: or just `Automattic\Jetpack\Assets`

File: modules/contact-form/grunion-form-view.php [hunk 2]
```diff
@@ -14,9 +17,12 @@
  */
 $max_new_fields = apply_filters( 'grunion_max_new_fields', 5 );
```
Comment (lezama) [inline modules/contact-form/grunion-form-view.php]: is this in the global scope? 😱

File: modules/contact-form/grunion-form-view.php [hunk 3]
```diff
@@ -14,9 +17,12 @@
...
 
+if ( ! $asset_tools instanceof Asset_Tools ) {
+	$asset_tools = new Asset_Tools();
```
Comment (lezama) [inline modules/contact-form/grunion-form-view.php]: this looks weird. I wonder if we should share via the global scope `$jetpack_get_file_url_for_environment` or `$jetpack_asset_tools`

File: class.jetpack-connection-banner.php (hunk 1 above)
Comment (lezama) [reply class.jetpack-connection-banner.php]: @gititon let's go `Automattic\Jetpack\Assets` ! seems to be the preference around the team.

**Review by gititon — COMMENTED** [review]
Thanks for the feedback here and on Slack. I'm working on implementing changes.

**Review by tyxla — COMMENTED** [review]
I know we've been through these discussions a bunch of times, but...

I wonder what's the benefit of using a singleton here...

* `Assets` object doesn't hook anything on `__construct`
* `Assets` objects don't contain any important state we need to preserve globally
* PHP will destruct any objects when they're no longer in use

So could we just initialize a new object and use it accordingly? Soon we'll be using PHP 5.6 only, and the `(new Foo)->bar()` syntax will be available, making this even shorter.

Is there any good reason that would make us prefer the singleton approach instead?

**Comment by gititon** [comment]
For context, this _is_ the refactor from instantiating a new object everywhere before using it.

**Comment by gititon** [comment]
Moving this back to Needs Review since it really needs a decision before any changes can be made.

**Comment by gravityrail** [comment]
Ok, I'm gonna go out on a limb here and suggest something a little different.

Currently we have `Jetpack::connection()` as a shortcut to a single instance of the connection. However, it has a dependency on a class which exists in the plugin but not a package. In theory we could introduce the `Jetpack` class into a `jetpack-core` package, but that feels a little... premature.

That said, how else can we avoid scattering injection of instances of `Assets` into the many places that use it, without using static methods? Static methods make it hard to make assertions on messages passed between class instances, and hard to mock and stub instances and methods.

I think the answer is to hold our noses and create a new pattern, which I would call the "overridable memoized instance" pattern (or something like that)

```php
// regular pattern
$url = Assets::instance()->get_file_url( 'foo.js' );

// test pattern - set memoized instance in `setUp()`
$my_mock_assets = $this->createMock(); 
Assets::instance( $my_mock_assets );
// now your regular code will receive the mock instance wherever it's called
```

Within the `Assets` class, you would do something like:

```php
class Assets {
  private static $instance;
  static function instance( $new_instance = null ) {
    if ( $new_instance ) {
      self::$instance = $new_instance;
    }

    if ( ! self::$instance ) {
      self::$instance = new Assets();
    }

    return self::$instance;
  }
}
```

Note that everything above is pseudocode, I haven't run any of it, and this is still open for debate. I'm just trying to offer my 2c in case it's helpful.

**Comment by gravityrail** [comment]
You could also use this to delegate static methods to instance methods:

```php
class Assets {
  public static function __callStatic($name, $arguments)
    {
        call_user_func_array( [ self::instance(), $name ], $arguments );
    }
}
```

This allows you to do something like:

```php
$url = Assets::get_file_url( 'foo.js' );
```

Which would be automatically delegated to `Assets::instance()->get_file_url( 'foo.js' )`

Then, in your test, you can still just do a setUp command like `Assets::instance( $my_mock_assets )` and your special instance will receive the delegated static calls, without you having to write ANY extra code to support it.

In theory you could put this in a `Utility_Singleton` base class which would allow us to make statically callable functions all over the place.

Cool huh?

**Comment by gititon** [comment]
This is on hold, and may be closed, in light of Dan's experimentation, https://github.com/Automattic/jetpack/pull/12744, which is itself on hold.

**Comment by gititon** [comment]
Closing in favor of https://github.com/Automattic/jetpack/pull/12744

**Comment by gititon** [comment]
https://github.com/Automattic/jetpack/pull/12744

**Comment by Bayuadi2401** [comment]
> This PR refactors Jetpack to use the Assets package for calling `get_file_url_for_environment` rather than the static method on the Jetpack class.
> 
> Since this is a refactor that should not introduce new functionality, please test by ensuring tests pass and by accessing various Jetpack pages/features and ensuring there are no errors in your error log.
> 
> In this latest revision, I used constructor dependency injection to inject Assets into a class when I could, and used a static get_instance() when it was needed from a static method.

use-asset-tools-package-everywhere

**Comment by Bayuadi2401** [comment]
```javascript
https://github.com/Automattic/jetpack/pull/12688#issuecomment-2181178326
```

**Pull request closed without merging**